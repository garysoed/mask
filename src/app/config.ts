import { VineImpl } from 'grapevine/export/main';
import { CustomElementCtrl } from 'persona/export/main';

export interface Config {
  /** TODO: Delete ctrl */
  ctor?: typeof CustomElementCtrl;
  dependencies?: Config[];
  tag: string;
  configure?(vine: VineImpl): void;
}
