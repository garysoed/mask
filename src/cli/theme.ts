import * as fs from 'fs';

import commandLineArgs from 'command-line-args';
import {Color, fromCssColor} from 'gs-tools/export/color';
import {EMPTY, Observable, Subject} from 'rxjs';
import {catchError, takeUntil} from 'rxjs/operators';
import {Logger, ON_LOG_$} from 'santa';
import {CliDestination} from 'santa/export/cli';

import {Palette, PALETTE} from '../theme/palette';
import {Theme} from '../theme/theme';


const LOGGER = new Logger('@hive/main');

const OPTIONS = [
  {
    name: 'base',
    type: String,
  },
  {
    name: 'acc',
    type: String,
  },
  {
    name: 'out',
    type: String,
  },
];

const options = commandLineArgs(OPTIONS, {stopAtFirstUnknown: true});

/**
 * Completes whenever running is done.
 */
function run(): Observable<unknown> {
  const base = findColor(options.base);
  const acc = findColor(options.acc);
  const outStr = options.out;
  const theme = new Theme(base, acc);

  return new Observable(subscriber => {
    fs.writeFile(outStr, theme.generateCss(), {encoding: 'utf8'}, err => {
      if (err) {
        subscriber.error(err);
      } else {
        subscriber.next();
        subscriber.complete();
      }
    });
  });
}

function findColor(colorStr: string): Color {
  const existingColor: Color|undefined = PALETTE[colorStr.toUpperCase() as keyof Palette];
  if (existingColor) {
    return existingColor;
  }

  const cssColor = fromCssColor(colorStr);
  if (!cssColor) {
    throw new Error(`Unsupported color: ${colorStr}`);
  }

  return cssColor;
}

const onDone$ = new Subject();
const consoleLog = new CliDestination(
    entry => {
      if (entry.key === 'authurl') {
        return {showKey: false, enableFormat: false};
      }
      return {showKey: false};
    },
);
ON_LOG_$
    .pipe(takeUntil(onDone$))
    .subscribe(entry => {
      consoleLog.log(entry);
    });

run()
    .pipe(
        catchError(e => {
          LOGGER.error(e);
          return EMPTY;
        }),
    )
    .subscribe({complete: () => {
      onDone$.next();
    }});
