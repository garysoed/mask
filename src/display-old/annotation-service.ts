import { source, stream, Vine } from 'grapevine';
import { map } from 'rxjs/operators';

import { AnnotationSpec } from './annotation-spec';


class AnnotationService {
  constructor(private readonly annotations: ReadonlyMap<string, AnnotationSpec>) { }

  getSpec(annotationId: string): AnnotationSpec|null {
    return this.annotations.get(annotationId) || null;
  }
}

const $annotationConfig = source('annotationConfig', () => new Map<string, AnnotationSpec>());

export function addAnnotationSpec(vine: Vine, annotationId: string, spec: AnnotationSpec): void {
  $annotationConfig.set(vine, configMap => new Map([...configMap, [annotationId, spec]]));
}

export const $annotationService = stream(
    'AnnotationService',
    vine => $annotationConfig.get(vine)
        .pipe(map(config => new AnnotationService(config))),
);
