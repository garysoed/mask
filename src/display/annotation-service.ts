import { source } from 'grapevine';
import { NodeWithId, PersonaContext } from 'persona';
import { Observable } from 'rxjs';

export type AnnotationSpec = (
    node: NodeWithId<Node>,
    context: PersonaContext,
) => Observable<ReadonlyArray<NodeWithId<Node>>>;

export type AnnotationConfig = ReadonlyMap<string, AnnotationSpec>;

export const $annotationConfig = source<AnnotationConfig>(
    'annotationConfig',
    () => new Map<string, AnnotationSpec>(),
);
