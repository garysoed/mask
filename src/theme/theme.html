<style>
  tr > td:first-child {
    padding-right: 2rem;
    text-align: end;
  }

  #col1 {
    width: 16rem;
  }

  #col2, #col3, #col4, #col5, #col6, #col7 {
    width: 12rem;
  }

  tbody tr {
    height: 8em;
  }

  thead th {
    text-align: center;
  }

  [mk-action-1].focus {
    --mkThemeBG: var(--mkThemeFocusBG1);
    --mkThemeFG: var(--mkThemeFocusFG1);
  }

  [mk-action-2].focus {
    --mkThemeBG: var(--mkThemeFocusBG2);
    --mkThemeFG: var(--mkThemeFocusFG2);
  }

  [mk-action-1].hover {
    --mkThemeBG: var(--mkThemeHoverBG1);
    --mkThemeFG: var(--mkThemeHoverFG1);
  }

  [mk-action-2].hover {
    --mkThemeBG: var(--mkThemeHoverBG2);
    --mkThemeFG: var(--mkThemeHoverFG2);
  }

  .label {
    font-size: 1rem;
  }

  .label.dark,
  #table.dark .label.light {
    display: none;
  }

  #table.dark .label.dark,
  .label.light {
    display: block;
  }
</style>

<table id="table" mk-theme-level="0">
  <colgroup>
    <col id="col1">
    <col id="col2">
    <col id="col3">
    <col id="col4">
    <col id="col5">
    <col id="col6">
    <col id="col7">
  </colgroup>
  <thead mk-body-2>
    <tr>
      <th></th>
      <th colspan="2">Main</th>
      <th colspan="2">Side</th>
      <th colspan="2">Highlight</th>
    </tr>
  </thead>
  <tbody>
    <!-- Passive -->
    <tr id="passiveRow">
      <td mk-body-2>passive</td>
      <td mk-passive-1 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-passive-1 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-passive-1 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Action -->
    <tr id="actionRow">
      <td mk-body-2>action</td>
      <td mk-action-1 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Focus -->
    <tr id="focusRow">
      <td mk-body-2>action (focus)</td>
      <td mk-action-1 class="focus" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="focus" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="focus" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Hover -->
    <tr id="hoverRow">
      <td mk-body-2>action (hover)</td>
      <td mk-action-1 class="hover" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="hover" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="hover" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Disabled -->
    <tr id="disabledRow">
      <td mk-body-2>action (disabled)</td>
      <td mk-action-1 mk-disabled mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-disabled mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-disabled mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>
  </tbody>
</table>
