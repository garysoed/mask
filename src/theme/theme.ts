import { $asArray, $flat, $map, $pipe } from 'gs-tools/export/collect';
import { Color } from 'gs-tools/export/color';
import { cache } from 'gs-tools/export/data';

import generalCss from './general.css';
import { PRIMARY_DARK_HIGHLIGHT_SECTION_SPECS, PRIMARY_DARK_SECTION_SPECS, PRIMARY_LIGHT_HIGHLIGHT_SECTION_SPECS, PRIMARY_LIGHT_SECTION_SPECS, SECONDARY_DARK_HIGHLIGHT_SECTION_SPECS, SECONDARY_DARK_SECTION_SPECS, SECONDARY_LIGHT_HIGHLIGHT_SECTION_SPECS, SECONDARY_LIGHT_SECTION_SPECS, SectionSpec } from './section-spec';
import { ColorWithAlpha, createColor } from './shade';


export class Theme {
  constructor(
      private readonly document: Document,
      readonly baseColor: Color,
      readonly accentColor: Color,
  ) { }

  getStyleEl(): HTMLStyleElement {
    return this.createStyleEl();
  }

  setBaseColor(color: Color): Theme {
    return new Theme(this.document, color, this.accentColor);
  }

  setHighlightColor(color: Color): Theme {
    return new Theme(this.document, this.baseColor, color);
  }

  @cache()
  private createStyleEl(): HTMLStyleElement {
    const cssDeclarations = [
      this.generateSectionThemeCss(false),
      this.generateSectionThemeCss(true),
      this.generateHighlightThemeCss(),
    ];

    const styleEl = this.document.createElement('style');
    styleEl.innerHTML = `${cssDeclarations.join('\n\n')}\n${generalCss}`;
    return styleEl;
  }

  private generateHighlightThemeCss(): string {
    // Just use the light ones. The dark ones will be the same.
    const cssVariables = $pipe(
        [...PRIMARY_LIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_SECTION_SPECS],
        $map(spec => {
          const cssNormalFgVar = getSectionVariableName(spec, true);
          const cssNormalBgVar = getSectionVariableName(spec, false);

          const highlightSpec = {...spec, isHighlight: true};
          const cssHighlightFgVar = getSectionVariableName(highlightSpec, true);
          const cssHighlightBgVar = getSectionVariableName(highlightSpec, false);

          return [
            `${cssNormalBgVar}: var(${cssHighlightBgVar});`,
            `${cssNormalFgVar}: var(${cssHighlightFgVar});`,
          ];
        }),
        $flat(),
        $asArray(),
    )
    .join('\n');

    return `[mk-theme-highlight] { ${cssVariables} }`;
  }

  private generateSectionThemeCss(isDark: boolean): string {
    const normalSectionSpecs = isDark ?
        [...PRIMARY_DARK_SECTION_SPECS, ...SECONDARY_DARK_SECTION_SPECS] :
        [...PRIMARY_LIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_SECTION_SPECS];
    const highlightSectionSpecs = isDark ?
        [...PRIMARY_DARK_HIGHLIGHT_SECTION_SPECS, ...SECONDARY_DARK_HIGHLIGHT_SECTION_SPECS] :
        [...PRIMARY_LIGHT_HIGHLIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_HIGHLIGHT_SECTION_SPECS];

    const cssVariables = $pipe(
        [...normalSectionSpecs, ...highlightSectionSpecs],
        $map(spec => {
          const base = spec.isBase ? this.baseColor : this.accentColor;
          const cssFgVar = getSectionVariableName(spec, true);
          const cssBgVar = getSectionVariableName(spec, false);
          const cssFgColor = generateCssColor(createColor(spec.fg, base));
          const cssBgColor = generateCssColor(createColor(spec.bg, base));
          return [
            `${cssFgVar}: ${cssFgColor};`,
            `${cssBgVar}: ${cssBgColor};`,
          ];
        }),
        $flat(),
        $asArray(),
    )
    .join('\n');

    const theme = isDark ? 'dark' : 'light';
    return `[mk-theme="${theme}"] { ${cssVariables} }`;
  }
}

function getSectionVariableName(sectionSpec: SectionSpec, isForeground: boolean): string {
  return [
    '--mkTheme',
    sectionSpec.section,
    isForeground ? 'FG' : 'BG',
    sectionSpec.isSecondary ? '2' : '1',
    sectionSpec.isHighlight ? 'H' : 'B',
  ].join('');
}

function generateCssColor({color, alpha}: ColorWithAlpha): string {
  return `rgba(${color.red},${color.green},${color.blue},${alpha})`;
}
