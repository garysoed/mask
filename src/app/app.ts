import { staticSourceId } from 'grapevine/export/component';
import { getOrRegisterApp as getOrRegisterGrapevineApp, VineImpl } from 'grapevine/export/main';
import { ImmutableMap } from 'gs-tools/export/collect';
import { Errors } from 'gs-tools/src/error';
import { InstanceofType } from 'gs-types/export';
import { CustomElementCtrl, getOrRegisterApp as getOrRegisterPersonaApp } from 'persona/export/main';
import { Palette } from '../theme/palette';
import { Theme } from '../theme/theme';
import { Config } from './config';

export const vine_ = getOrRegisterGrapevineApp('maskBase');
export const persona_ = getOrRegisterPersonaApp('maskBase', vine_);

export const $theme = staticSourceId('theme', InstanceofType(Theme));
vine_.builder.source($theme, new Theme(Palette.ORANGE, Palette.GREEN));

export function addToMapConfig_(map: Map<typeof CustomElementCtrl, Config>, config: Config): void {
  const existingConfig = map.get(config.ctor);
  if (existingConfig && existingConfig !== config) {
    throw Errors.assert(`Config for ${config.ctor.name}`).should('not be defined differently')
        .butNot();
  }

  map.set(config.ctor, config);
}

export function flattenConfigs_(configs: Config[]): ImmutableMap<typeof CustomElementCtrl, Config> {
  const map = new Map<typeof CustomElementCtrl, Config>();
  for (const config of configs) {
    addToMapConfig_(map, config);

    const dependencies = config.dependencies || [];
    const dependencyMap = flattenConfigs_(dependencies);
    for (const [, config] of dependencyMap) {
      addToMapConfig_(map, config);
    }
  }

  return ImmutableMap.of(map);
}

export function start(
    configs: Config[],
    theme: Theme,
    styleEl: HTMLStyleElement,
    customElementRegistry: CustomElementRegistry = window.customElements): {vine: VineImpl} {
  const flattenedConfigs = flattenConfigs_(configs);

  const ctors = [];
  for (const [ctor] of flattenedConfigs) {
    ctors.push(ctor);
  }

  const vine = vine_.builder.run();
  vine.setValue($theme, theme);
  persona_.builder.build(ctors, customElementRegistry, vine);

  for (const [, {configure}] of flattenedConfigs) {
    if (configure) {
      configure(vine);
    }
  }

  vine.listen(theme => theme.injectCss(styleEl), $theme);

  return {vine};
}
