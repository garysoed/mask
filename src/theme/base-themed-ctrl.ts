import {BaseCtrl, PersonaContext} from 'persona';
import {Observable} from 'rxjs';
import {map, pairwise, startWith, tap} from 'rxjs/operators';

import {$themeLoader, _p} from '../app/app';


@_p.baseCustomElement({})
export abstract class BaseThemedCtrl<S extends {}> extends BaseCtrl<S> {
  protected readonly themeLoader$ = $themeLoader.get(this.vine);

  constructor(context: PersonaContext, specs: S) {
    super(context, specs);

    this.addSetup(this.setupThemeUpdate());
  }

  private setupThemeUpdate(): Observable<unknown> {
    // TODO: Replace with Persona's single.
    return this.themeLoader$.pipe(
        map(themeLoader => themeLoader.createElement(this.shadowRoot.ownerDocument)),
        startWith(null),
        pairwise(),
        tap(([oldEl, newEl]) => {
          if (oldEl) {
            oldEl.remove();
          }

          if (newEl) {
            this.shadowRoot.appendChild(newEl);
          }
        }),
    );
  }
}
