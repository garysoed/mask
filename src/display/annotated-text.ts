import {$asArray, $filterNonNull, $map, $pipe} from 'gs-tools/export/collect';
import {cache} from 'gs-tools/export/data';
import {debug} from 'gs-tools/export/rxjs';
import {attributeIn, host, listParser, multi, NodeWithId, PersonaContext, root, setId, stringParser, textIn, ValuesOf} from 'persona';
import {combineLatest, concat, Observable, of as observableOf} from 'rxjs';
import {bufferCount, map, switchMap, withLatestFrom} from 'rxjs/operators';
import {Logger} from 'santa';

import {_p} from '../app/app';
import {BaseThemedCtrl} from '../theme/base-themed-ctrl';

import {$annotationConfig, AnnotationSpec} from './annotation-service';


const LOGGER = new Logger('mask.AnnotatedText');

export const $annotatedText = {
  tag: 'mk-annotated-text',
  api: {
    annotations: attributeIn('annotations', listParser(stringParser()), []),
    text: textIn(),
  },
};

export const $ = {
  host: host($annotatedText.api),
  root: root({
    content: multi('#contents'),
  }),
};

@_p.customElement({
  ...$annotatedText,
  template: '<!-- #contents -->',
})
export class AnnotatedText extends BaseThemedCtrl<typeof $> {
  constructor(context: PersonaContext) {
    super(context, $);
  }

  @cache()
  protected get values(): ValuesOf<typeof $> {
    return {
      root: {
        content: this.content$,
      },
    };
  }

  @cache()
  private get annotations$(): Observable<readonly AnnotationSpec[]> {
    return this.inputs.host.annotations.pipe(
        map(annotations => annotations || []),
        withLatestFrom($annotationConfig.get(this.vine)),
        map(([annotationIds, config]) => $pipe(
            annotationIds,
            $map(annotationId => config.get(annotationId) ?? null),
            $filterNonNull(),
            $asArray(),
        )),
    );
  }

  private applyAnnotation(
      nodes: ReadonlyArray<NodeWithId<Node>>,
      spec: AnnotationSpec,
  ): Observable<ReadonlyArray<NodeWithId<Node>>> {
    return concat(...nodes.map(node => spec(node, this.context))).pipe(
        bufferCount(nodes.length),
        map(outputs => {
          return outputs.reduce((acc, curr) => [...acc, ...curr]);
        }),
    );
  }

  private applyAnnotations(
      node: Node,
      specs: readonly AnnotationSpec[],
  ): Observable<ReadonlyArray<NodeWithId<Node>>> {
    if (specs.length <= 0) {
      return observableOf([setId(node, node)]);
    }

    let obs: Observable<ReadonlyArray<NodeWithId<Node>>> = observableOf([setId(node, node)]);
    for (const spec of specs) {
      obs = obs.pipe(switchMap(nodes => this.applyAnnotation(nodes, spec)));
    }

    return obs;
  }

  @cache()
  private get content$(): Observable<ReadonlyArray<NodeWithId<Node>>> {
    return combineLatest([
      this.inputs.host.text,
      this.annotations$,
    ])
        .pipe(
            switchMap(([text, annotations]) => {
              return this.applyAnnotations(document.createTextNode(text), annotations);
            }),
            debug(LOGGER, 'content'),
        );
  }
}
