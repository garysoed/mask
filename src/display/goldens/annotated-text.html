<mk-annotated-text mk-theme="light">banana<gst-fake-shadow><!-- #contents -->ccncnc<link rel="stylesheet" href="base/test.css"></gst-fake-shadow></mk-annotated-text>